### PR TITLE
New iterative dihedral fitting procedure

### PR DESCRIPTION
**Pull request by stefdoerr**
Linked to this data commit https://github.com/Acellera/htmd-data/pull/34

### CODE REVIEW

File: htmd/parameterization/dihedral.py [hunk 1]
```diff
@@ -330,41 +345,91 @@ def _evaluateConstTerms(self):
...
         if self.zeroed_parameters:
             vector[:] = 0
 
-        # Optimize the parameters
-        logger.info('Start parameter optimization')
+        # Optimize each dihedral individually
+        logger.info('Start parameter optimization with {} iterations'.format(self.num_iterations))
         start = time.clock()
-        vector = self._optimizeWithRandomSearch(vector)
+        best_loss = None
```
Comment (raimis) [inline htmd/parameterization/dihedral.py]: This should be factor out into `self._optimizeWithXXX(vector)`

File: htmd/parameterization/dihedral.py [hunk 2]
```diff
@@ -330,41 +345,91 @@ def _evaluateConstTerms(self):
...
+
+            # Fit the individual dihedrals
+            dihedral_vectors = []
+            for idihed in range(self.numDihedrals):
+                logger.info('Optimizing dihedral {}/{}'.format(idihed+1, self.numDihedrals))
+                subvector = self._paramsToVector(current_params, idihed)
+                subvector = self._optimizeWithRandomSearch(subvector, target_energies_dihedral, num_searches=20, idihed=idihed, _logger=False)
+                dihedral_vectors.append(subvector)
+
+            # Pack the individually fitted parameters into a vector
```
Comment (raimis) [inline htmd/parameterization/dihedral.py]: This could be factored out to `self._subvectorToVector` or something like that.

File: htmd/parameterization/cli.py [hunk 3]
```diff
@@ -51,8 +51,8 @@ def getArgumentParser():
                         help='DEPRECATED: use `--scan-type` instead')
     parser.add_argument('--scan-type', default='qm', dest='dihed_opt_type', choices=['None', 'qm', 'mm'],
                         help='Type of structure optimization when scanning dihedral angles (default: %(default)s)')
-    parser.add_argument('--dihed-num-searches', default=None, type=int,
-                        help='Number of random searches during the dihedral parameter fitting')
+    parser.add_argument('--dihed-num-iterations', default=3, type=int,
```
[the remark this answers is not in the document]
Comment (stefdoerr) [reply htmd/parameterization/cli.py]: Done

File: htmd/parameterization/dihedral.py [hunk 4]
```diff
@@ -26,6 +26,20 @@
 logger = logging.getLogger(__name__)
 
 
+def _evaluateScanEnergies(k0, phi0, offset, n, angle_values, numScans, dihedrals):
```
[the remark this answers is not in the document]
Comment (stefdoerr) [reply htmd/parameterization/dihedral.py]: I moved it closer to the gradients if it's a matter of distance in lines of code. Tests will catch the bug

File: htmd/parameterization/dihedral.py (hunk 1 above)
Comment (stefdoerr) [reply htmd/parameterization/dihedral.py]: Done

File: htmd/parameterization/dihedral.py (hunk 2 above)
Comment (stefdoerr) [reply htmd/parameterization/dihedral.py]: Done

File: htmd/parameterization/dihedral.py [hunk 5]
```diff
@@ -242,39 +253,72 @@ def _vectorToParams(self, vector):
 
         return parameters
 
-    def _optimizeWithRandomSearch(self, vector):
+    def _optimizeWithIterativeScheme(self, vector, target_energies):
```
Comment (raimis) [inline htmd/parameterization/dihedral.py]: This is getting a spaghetti code. i.e `self._optimizeWithRandomSearch` should not be coerced to do local optimizations and other stuff not related wih NRS.

Just replicate the creation of local optimizers (could be factored out in the future) and use them directly. Random sampling is also just a few lines of code.

File: htmd/parameterization/cli.py [hunk 6]
```diff
@@ -51,8 +51,10 @@ def getArgumentParser():
...
```
[the remark this answers is not in the document]
Comment (stefdoerr) [reply htmd/parameterization/cli.py]: No that's not true. NRS will do the number of iterations it computes. It correctly ignores the num_iterations.

File: htmd/parameterization/dihedral.py [hunk 7]
```diff
@@ -154,68 +151,82 @@ def _getBounds(self):
...
 
-    def _objective(self, x, grad=None):
+    def _unpackVector(self, vector):
+        k0, phi0 = np.reshape(vector[:-1], (2, -1, self.MAX_DIHEDRAL_MULTIPLICITY))
+        offset = vector[-1]
+        n = np.arange(1, self.MAX_DIHEDRAL_MULTIPLICITY + 1)
+        return k0, phi0, offset, n
+
+    @staticmethod
+    def _evaluateScanEnergies(k0, phi0, offset, n, angle_values, numScans, dihedrals):
```
[the remark this answers is not in the document]
Comment (stefdoerr) [reply htmd/parameterization/dihedral.py]: I am against that coding style. It creates confusion as to who computes what values and hides the flow of information. 
The only variables that should be stored in a class should be "state" and option variables. As these variables don't reflect the state of the class but the state of a computation of a class I do not store them.

File: htmd/parameterization/dihedral.py [hunk 8]
```diff
@@ -242,39 +253,72 @@ def _vectorToParams(self, vector):
...
+                                  num_searches=None, best_vector=None, best_loss=None, _logger=True):
         """
         Naive random search
         """
+        numDihedrals = self.numDihedrals if idihed is None else 1
 
         # Create a local optimizer
         opt = nlopt.opt(nlopt.LD_LBFGS, vector.size)
         opt.set_vector_storage(opt.get_dimension())
-        logger.info('Local optimizer: {} with {} vector storage'.format(opt.get_algorithm_name(),
```
[the remark this answers is not in the document]
Comment (stefdoerr) [reply htmd/parameterization/dihedral.py]: It's information that tells nothing to the user. He cannot change it. Also it gets printed at every individual dihedral optimization which makes for a bit of spam.

File: htmd/parameterization/cli.py [hunk 9]
```diff
@@ -51,8 +51,10 @@ def getArgumentParser():
...
+                        help='Number of iterations during the dihedral parameter fitting')
+    parser.add_argument('--dihed-fit-type', default='Iterative', choices=['Iterative', 'NRS'],
```
[the remark this answers is not in the document]
Comment (stefdoerr) [reply htmd/parameterization/cli.py]: Why should the default be NRS? It was quite clear to me that we chose to use the iterative scheme as the default. We will even add it to the paper

File: htmd/parameterization/dihedral.py [hunk 10]
```diff
@@ -330,41 +377,73 @@ def _evaluateConstTerms(self):
 
         return const_energies
 
-    def _fit(self):
+    def _evaluateConstTermsPerDihedral(self, parameters):
+        # for key in parameters.dihedral_types: print(key, [term.phi_k for term in parameters.dihedral_types[key]])
```
[the remark this answers is not in the document]
Comment (stefdoerr) [reply htmd/parameterization/dihedral.py]: Done

File: htmd/parameterization/dihedral.py [hunk 11]
```diff
@@ -154,68 +151,82 @@ def _getBounds(self):
...
+
+    @staticmethod
+    def _evaluateScanEnergies(k0, phi0, offset, n, angle_values, numScans, dihedrals):
+        actual_energies = []
+
+        for iscan in range(numScans):
+            current_energy = 0
+            for i, idihed in enumerate(dihedrals):
+                phis = angle_values[iscan][idihed][:, :, None]
+                energies = np.sum(k0[i] * (1 + np.cos(n * phis - phi0[i])), axis=(1, 2)) + offset
```
[the remark this answers is not in the document]
Comment (stefdoerr) [reply htmd/parameterization/dihedral.py]: Fixed

File: htmd/parameterization/dihedral.py [hunk 12]
```diff
@@ -154,68 +151,82 @@ def _getBounds(self):
...
+        offset = vector[-1]
+        n = np.arange(1, self.MAX_DIHEDRAL_MULTIPLICITY + 1)
+        return k0, phi0, offset, n
+
+    @staticmethod
+    def _evaluateScanEnergies(k0, phi0, offset, n, angle_values, numScans, dihedrals):
+        actual_energies = []
+
+        for iscan in range(numScans):
+            current_energy = 0
```
[the remark this answers is not in the document]
Comment (stefdoerr) [reply htmd/parameterization/dihedral.py]: Fixed

File: htmd/parameterization/dihedral.py [hunk 13]
```diff
@@ -97,11 +94,11 @@ def _setup(self):
         all_equivalent_dihedrals = {tuple(dihedrals[0]): dihedrals for dihedrals in all_equivalent_dihedrals}
 
         # Choose the selected dihedrals
-        self._equivalent_dihedrals = []
+        _equivalent_dihedrals = []
```
[the remark this answers is not in the document]
Comment (stefdoerr) [reply htmd/parameterization/dihedral.py]: Leftover from refactoring out the `self.`. Fixed now

File: htmd/parameterization/dihedral.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (stefdoerr) [reply htmd/parameterization/dihedral.py]: I added back in information on the number of random searches as it might be useful to know how long it will take. The optimizer information I still don't see as of any use as I said before. We are not planning on making the NRS default anyway so it doesn't matter too much.

File: htmd/parameterization/dihedral.py (hunk 5 above)
Comment (stefdoerr) [reply htmd/parameterization/dihedral.py]: I made the Iterative scheme now optimize directly instead of setting the num_searches to 1. I refactored out the two optimizers which also avoids the previous re-initialization of optimizers.

File: htmd/parameterization/dihedral.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (raimis) [reply htmd/parameterization/dihedral.py]: I meant `angle_values` and `numScans`, which doesn't change.
